infoschema/builder.go

```diff
@@ -367,6 +369,38 @@ func (b *Builder) applyDropSchema(schemaID int64) []int64 {
 	return tableIDs
 }
 
+func (b *Builder) applyRecoverSchema(m *meta.Meta, diff *model.SchemaDiff) error {
+	if di, ok := b.is.SchemaByID(diff.SchemaID); ok {
+		return ErrDatabaseExists.GenWithStackByArgs(
+			fmt.Sprintf("(Schema ID %d)", di.ID),
+		)
+	}
+	di, err := m.GetDatabase(diff.SchemaID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err = b.applyPlacementUpdate(placement.GroupID(diff.SchemaID)); err != nil {
+		return errors.Trace(err)
+	}
+	tables, err := m.ListTables(diff.SchemaID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	var tids []int64
+	b.is.schemaMap[di.Name.L] = &schemaTables{
+		dbInfo: di,
+		tables: make(map[string]table.Table, len(tables)),
+	}
+	for _, tblInfo := range tables {
+		allocs := autoid.NewAllocatorsFromTblInfo(b.store, di.ID, tblInfo)
+		tids, err = b.applyCreateTable(m, di, tblInfo.ID, allocs, diff.Type, tids)
+		if err != nil {
+			return errors.Trace(err)
+		}
```

Comment:
also need `applyPlacementUpdate` for each table?

Reply:
@erwadba Please check this comment.

Reply:
@xhebox Could you give some advice?